pkg/api/v1/types.go

```diff
@@ -1217,12 +1217,14 @@ type EndpointSubset struct {
 	Ports     []EndpointPort    `json:"ports,omitempty" description:"port numbers available on the related IP addresses"`
 }
 
-// EndpointAddress is a tuple that describes single IP address.
+// EndpointAddress is a tuple that describes single endpoint address.
 type EndpointAddress struct {
-	// The IP of this endpoint.
-	// TODO: This should allow hostname or IP, see #4447.
+	// The IP of this endpoint, mutually exclusive with Hostname
 	IP string `json:"ip" description:"IP address of the endpoint"`
 
+	// Hostname for this endpoint, you can specify IP or Hostname, but not both.
+	Hostname string `json:"hostname" description:"Hostname for this endpoint"`
```

Comment:
I can sorta see what you're trying to do here, but the details seem very murky. Do you anticipate the endpoints controller ever setting this field? If so, how? Don't we need to solve external DNS before we'll have something reasonable to put here? If not, should this really go here?

Is Hostname the right field name? How about "DNSName" or "DNS"?

Why not combine this with the IP field? Do we really need two fields?